numpy/core/arrayprint.py

```diff
@@ -1596,12 +1597,98 @@ def array_str(a, max_line_width=None, precision=None, suppress_small=None):
         a, max_line_width, precision, suppress_small)
 
 
+def _parse_format_spec(fs):
+    """
+    Parse the format spec and returns a dictionary with options for using it with `array2string`
+    This is inspired in the [Format Specification Mini-Language](https://docs.python.org/3/library/string.html#formatspec)
+
+    Parameters
+    ----------
+    fs: str
+        a string with the format specification, e.g., "+.2f"
+
+    Semantic for format spec
+    ------------------------
+
+    format_spec ::=  [sign][.precision][type]
+    sign        ::=  "+" | "-" | " "
+    precision   ::=  [0-9]+
+    type        ::=  "f" | "e"
+    """
+
+    options = {}
+    fslen = len(fs)
+    pos = 0
+
+    # Parse the sign options
+    if fslen - pos >= 1 and fs[pos] in ["+", "-", " "]:
+        options["sign"] = fs[pos]
+
+        pos += 1
+
+    # Parse the field precision
+    if fslen - pos >= 1 and fs[pos] == ".":
+        pos += 1
+
+        posi = pos
+        while pos < fslen and str.isdigit(fs[pos]):
+            pos += 1
+
+        if posi == pos:  # Format specifier missing precision
+            raise ValueError("Format specifier missing precision")
+
+        options["precision"] = int(fs[posi:pos])
+
+    # Finally, parse the field type
+
+    if fslen - pos > 1:  # more than 1 char remain, invalid format specifier
+        raise ValueError(f"Invalid format specifier '{fs[pos]}'")
+
+    if fslen - pos == 1:
+        if fs[pos] == "f":  # force fixed-point notation
+            options["suppress_small"] = True
+        elif fs[pos] == "e":  # use scientific notation when necessary
+            options["suppress_small"] = False
+        else:
+            raise ValueError(f"Unknown format code '{fs[pos]}'")
+
+    return options
+
+
+def _array_format_implementation(a, format_spec):
+    if issubclass(a.dtype.type, np.object_):
+        # this will raise TypeError if format_spec is not an empty string
+        # this follow what list() or tuple() does in Python
+        return object.__format__(a, format_spec)
+
+    # default behaviour when trying to format non-numeric dtypes
+    if not issubclass(a.dtype.type, np.number):
+        raise TypeError(f"Format spec. not implemented for dtype={a.dtype}")
+
+    options = _parse_format_spec(format_spec)
+
+    return array2string(a, **options)
+
+
+def _array_format_dispatcher(
+        a, format_spec):
+    return (a, format_spec)
+
+
+@array_function_dispatch(_array_format_dispatcher, module='numpy')
+def array_format(a, format_spec):
+    print("[DEBUG] array_format")
+
+    return _array_format_implementation(a, format_spec)
```

Comment:
Why do we need to expose a new public function here?
Won't `builtins.format` suffice for public usage?

Reply:
I agree — "there should only be one way to do it", and leaving this name private will encourage folks to encounter this through the ecosystem of f'', .format(), and builtins.format(), rather than through a bespoke mechanism.

Reply:
Actually, you have a very good point. I really didn't think about this and I tried to do it the same way as it's done with `array2string`. Do you know if there is any other use for the `array_function_dispatch` decorator apart from exposing the function with another name?

Reply:
I don't see why we have need `array_function_dispatch` here in the first place. The relevant `format` output can already be fully customized via custom implementations of `__format__`.

Reply:
In the code, there are several examples in which they use it and I can't seem to understand them: `array_str` and `array_repr` instead of using plain `str()` and `repr()`. Is there any use for that?

Reply:
> In the code, there are several examples in which they use it and I can't seem to understand them: `array_str` and `array_repr` instead of using plain `str()` and `repr()`. Is there any use for that?

Considering all these functions are at least 17 years old, I suspect it is mostly due to "historical reasons". NEP 18 came quite a bit latter and by that time support was added to `array_str` and `array_repr` as they were already well established.

Reply:
This function is still exposed in the namespace. Either remove it or start a process to expose it: add a proper docstring and post the justification to the mailing list.